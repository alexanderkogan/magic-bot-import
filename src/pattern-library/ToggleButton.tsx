import * as React from 'react'
import { Props } from '../react-typings'

export interface ToggleButtonProps<T extends string = string> {
    id: string
    values: T[]
    changed: (newValue: T) => void
    currentValue?: T
}

function createInput<T extends string = string>(id: string, value: T, props: ToggleButtonProps<T>): JSX.Element {
    return (
        <input
            type="radio"
            name="radio-group"
            id={id}
            key={`input_${id}`}
            onChange={() => props.changed(value)}
            checked={props.currentValue === value}
        />
    )
}

function createLabel(id: string, value: string): JSX.Element {
    return (
        <label key={`label_${id}`} htmlFor={id}>
            {value}
        </label>
    )
}

export default <T extends string = string>(props: Props<ToggleButtonProps<T>>) => {
    const toggles: JSX.Element[] = []
    props.values.forEach((value, index) => {
        const id = `${props.id}${index}`
        const input = createInput(id, value, props)
        const label = createLabel(id, value)
        // we have to render inputs and labels alternating without a div wrapper. Jsx does not allow this directly,
        // so we collect them in an array.
        toggles.push(input, label)
    })
    return <div>{toggles}</div>
}
