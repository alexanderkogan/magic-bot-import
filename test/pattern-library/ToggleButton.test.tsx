import * as React from 'react'

import ToggleButton from '../../src/pattern-library/ToggleButton'
import { mount } from '../test-utils'

describe('ToggleButton', () => {
    it('renders', () => {
        const element = mount(<ToggleButton id="toggle1" values={['yes', 'maybe', 'no']} changed={jest.fn} currentValue="yes" />)
        expect(element).toMatchSnapshot()
    })

    it('calls changed on change event', () => {
        const changed = jest.fn()
        const element = mount(<ToggleButton id="toggle1" values={['yes', 'maybe', 'no']} changed={changed} currentValue="yes" />)
        element
            .find('input')
            .at(2)
            .simulate('change')
        expect(changed).toHaveBeenCalledWith('no')
    })
})
