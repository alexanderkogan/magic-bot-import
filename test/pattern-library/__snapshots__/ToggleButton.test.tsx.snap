// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton renders 1`] = `
<Component
  changed={[Function]}
  currentValue="yes"
  id="toggle1"
  values={
    Array [
      "yes",
      "maybe",
      "no",
    ]
  }
>
  <div>
    <input
      checked={true}
      className="obc_btn-toggle"
      id="toggle10"
      key="input_toggle10"
      name="radio-group"
      onChange={[Function]}
      type="radio"
    />
    <label
      htmlFor="toggle10"
      key="label_toggle10"
    >
      yes
    </label>
    <input
      checked={false}
      className="obc_btn-toggle"
      id="toggle11"
      key="input_toggle11"
      name="radio-group"
      onChange={[Function]}
      type="radio"
    />
    <label
      htmlFor="toggle11"
      key="label_toggle11"
    >
      maybe
    </label>
    <input
      checked={false}
      className="obc_btn-toggle"
      id="toggle12"
      key="input_toggle12"
      name="radio-group"
      onChange={[Function]}
      type="radio"
    />
    <label
      htmlFor="toggle12"
      key="label_toggle12"
    >
      no
    </label>
  </div>
</Component>
`;
